server/src/citypay/citypay.service.ts

```diff
@@ -1,120 +1,135 @@
+        <quantity>1</quantity>
+        <sequenceNumber>${i}</sequenceNumber>
+        <unitPrice>${curInvoice.dcp_grandtotal}</unitPrice>
+      </retailPaymentRequestLineItems>`)
     }
-  }
 
-  // generates cookies by headless login to MS Applicant Portal
-  private async stealCookies(retries = MAX_RETRIES) {
-    const browserPromise = puppeteer.launch({ args: ['--no-sandbox'] });
+    return `<RetailPaymentRequest xmlns="${this.config.get('CITYPAY_DOMAIN')}">
+          <profileID>${this.config.get('CITYPAY_USERNAME')}</profileID>
+          <profilePassword>${this.config.get('CITYPAY_PASSWORD')}</profilePassword>
+          <displayTransactionDescription_1></displayTransactionDescription_1>
+          <displayTransactionDescription_2></displayTransactionDescription_2>
+          <displayTransactionDescription_3></displayTransactionDescription_3>
+          <paymentExtraData></paymentExtraData>
+          <postBackURL>${this.config.get('CITYPAY_POSTBACK')}</postBackURL>
+          <returnFromCartURL>${this.config.get('CITYPAY_RETURN_FROM_CART')}</returnFromCartURL>
+          <returnFromCheckoutURL>${this.config.get('CITYPAY_RETURN_FROM_CHECKOUT')}</returnFromCheckoutURL>
+          <agencyRequestID>${agencyRequestID}</agencyRequestID>
+          ${lineItems.join('')}
+        </RetailPaymentRequest>`;
+  }
 
-    try {
-      const browser = await browserPromise;
-      const page = await browser.newPage();
+  async getCartKey(packageId) {
+    const agencyRequestID = uuidv4();
 
-      await page.goto(MS_APPLICANT_PORTAL.SIGN_IN_PAGE);
+    const { records: [firstPackage] } = await this.crmService.get('dcp_packages', `
+        $filter=dcp_packageid eq ${packageId}
+        &$expand=dcp_dcp_package_dcp_projectinvoice_package(
+          $filter=statuscode eq ${DCP_PROJECTINVOICE_CODES.statuscode.APPROVED}
+        )`
+    );
 
-      // this gets dynamically inserted after a handshake event with MS
-      await page.waitForSelector('[action="/SignIn"]');
+    const requestXML = await this.createRequestXML(firstPackage, agencyRequestID);
 
-      // enter credentials
-      await page.type('#Username', this.config.get('MS_APPLICANT_PORTAL_USERNAME'));
+    const params = new url.URLSearchParams({ saleData: requestXML });
 
-      // Something changed MCS-portal side for this selector
-      // the selector changed from Password -> PasswordValue
-      // this code will try both just in case.
-      try {
-        await page.type('#Password', this.config.get('MS_APPLICANT_PORTAL_PASSWORD'));
-      } catch (e) {
-        await page.type('#PasswordValue', this.config.get('MS_APPLICANT_PORTAL_PASSWORD'));
-      }
+    try {
+      let citypayResponse = null;
 
-      // click signin, but await for the response with cookies
-      page.click('#submit-signin-local');
+      citypayResponse = await axios.post(`${this.config.get('PAYMENT_BASE_URL')}/${this.config.get('PAYMENT_STEP1_URL')}`, params.toString(), {
```

Comment:
Can you consolidate `PAYMENT_BASE_URL` AND `CITYPAY_CUSTOMER_LINK` into one variable? It looks like they are the same for most of the full path. Related, can you rename `PAYMENT_STEP1_URL`? I know that's what i was called in the C# code but that's not a very descriptive variable name.

Reply:
I'm going to test this in a future PR because currently I can't test locally due to the node bump